compiler/noirc_frontend/src/hir_def/types.rs

```diff
@@ -187,6 +187,26 @@ impl Type {
     }
 }
 
+/// Types need to be distinguished depending on the scenario in which they can be used.
+///
+/// For example, the type of a struct field or a function parameter is expected to be
+/// a normal type. While the type we expect in an array expression (`[<some value>; <numeric type>]`)
+/// is expected to be a numeric type.  
+#[derive(PartialEq, Eq, Clone, Copy, Hash, Debug)]
+pub enum TypeKind {
+    Normal,
+    Numeric,
```

Comment:
Should we include the expected numeric type here as well?

Reply:
I ended up adding the numeric type here with the refactor where I only add numeric generics into the function scope.